### ISSUE TITLE
1.0.2m broke win32 nmake build vs 1.0.2l due to ./util/copy-if-different.pl changes

### ISSUE THREAD

**Issue by dch** [openssl/openssl#4680]
## background

To build 32-bit Erlang/OTP on windows, for 32-bit Apache CouchDB, we use a simple `nmake` build of
OpenSSL running under command prompt, via script[2] unchanged in 5 years. I'm trying to use it to
build the latest 1.0.2m series unsuccessfully. 1.0.2l *does* work however.

[2]: https://github.com/apache/couchdb-glazier/blob/v1.6.1/bin/build_openssl.cmd 

## errors

```
perl .\util\copy-if-different.pl ".\.\e_os.h" "tmp32\e_os.h"
Can't Open "././e_os.h" at .\util\copy-if-different.pl line 67.
NMAKE : fatal error U1077: 'c:\relax\strawberry\perl\bin\perl.EXE' : return code '0x16'
Stop.

## and

c:\relax\openssl-1.0.2m\crypto\../crypto/constant_time_locl.h(50) : fatal error C1083: Cannot open include file: 'e_os.h': No such file or directory
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin\cl.EXE"' : return code '0x2'
Stop.
```

I see these scripts have been changed again in /master/ branch, but I'm not clear how I could backport
those changes into 1.0.2m to confirm if they do indeed resolve this issue - mk1f.pl and so forth are
all gone now, so it seems quite a big change.

## full log

```
C:\relax>build_openssl

C:\relax>setlocal

C:\relax>path=C:\Windows\Microsoft.NET\Framework\v4.0.30319;C:\Windows\Microsoft.NET\Framework\v3.5;;C:\Program Files (x86)\Microsoft Visual Studio 10.0\Common7\IDE;C:\Program Files (x86)\Microsoft Visual Studio 10.0\Common7\Tools;;C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin;C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin\VCPackages;;C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin\NETFX 4.0 Tools;C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin;;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files\Microsoft Windows Performance Toolkit\;C:\Program Files\nodejs\;C:\node\v0.8.25\x64\;C:\ProgramData\chocolatey\bin;C:\Program Files\Git\cmd;z:\dropbox\tools;z:\repos\apps\bin;c:\relax\bin;C:\Users\couchdb\AppData\Roaming\npm\;c:\mozilla-build\7zip;c:\relax\nasm;c:\relax\strawberry\perl\bin;

C:\relax>7z x "c:\TEMP\openssl-*.tar" -oc:\relax\ -y

7-Zip 4.42  Copyright (c) 1999-2006 Igor Pavlov  2006-05-14

Processing archive: c:\TEMP\openssl-1.0.2m.tar

Extracting  openssl-1.0.2m\ACKNOWLEDGMENTS
...
Extracting  openssl-1.0.2m\VMS\VMSify-conf.pl
Extracting  openssl-1.0.2m\VMS\WISHLIST.TXT

Everything is Ok

C:\relax>pushd c:\relax\openssl-1.0.2m

c:\relax\openssl-1.0.2m>perl Configure VC-WIN32 --prefix=c:\relax\openssl enable-static-engine
Configuring for VC-WIN32
    no-ec_nistp_64_gcc_128 [default]  OPENSSL_NO_EC_NISTP_64_GCC_128 (skip dir)
    no-gmp          [default]  OPENSSL_NO_GMP (skip dir)
    no-jpake        [experimental] OPENSSL_NO_JPAKE (skip dir)
    no-krb5         [krb5-flavor not specified] OPENSSL_NO_KRB5
    no-libunbound   [experimental] OPENSSL_NO_LIBUNBOUND (skip dir)
    no-md2          [default]  OPENSSL_NO_MD2 (skip dir)
    no-rc5          [default]  OPENSSL_NO_RC5 (skip dir)
    no-rfc3779      [default]  OPENSSL_NO_RFC3779 (skip dir)
    no-sctp         [default]  OPENSSL_NO_SCTP (skip dir)
    no-shared       [default]
    no-ssl-trace    [default]  OPENSSL_NO_SSL_TRACE (skip dir)
    no-ssl2         [default]  OPENSSL_NO_SSL2 (skip dir)
    no-store        [experimental] OPENSSL_NO_STORE (skip dir)
    no-unit-test    [default]  OPENSSL_NO_UNIT_TEST (skip dir)
    no-weak-ssl-ciphers [default]  OPENSSL_NO_WEAK_SSL_CIPHERS (skip dir)
    no-zlib         [default]
    no-zlib-dynamic [default]
IsMK1MF=1
CC            =cl
CFLAG         =-DOPENSSL_THREADS  -DDSO_WIN32 -W3 -Gs0 -GF -Gy -nologo -DOPENSSL_SYSNAME_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -D_CRT_SECURE_NO_DEPRECATE -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SS
E2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM
EX_LIBS       =
CPUID_OBJ     =x86cpuid.o
BN_ASM        =bn-586.o co-586.o x86-mont.o x86-gf2m.o
EC_ASM        =
DES_ENC       =des-586.o crypt586.o
AES_ENC       =aes-586.o vpaes-x86.o aesni-x86.o
BF_ENC        =bf-586.o
CAST_ENC      =cast-586.o
RC4_ENC       =rc4-586.o
RC5_ENC       =rc5-586.o
MD5_OBJ_ASM   =md5-586.o
SHA1_OBJ_ASM  =sha1-586.o sha256-586.o sha512-586.o
RMD160_OBJ_ASM=rmd-586.o
CMLL_ENC      =cmll-x86.o
MODES_OBJ     =ghash-x86.o
ENGINES_OBJ   =
PROCESSOR     =
RANLIB        =true
ARFLAGS       =
PERL          =perl
THIRTY_TWO_BIT mode
BN_LLONG mode
RC4_INDEX mode
RC4_CHUNK is undefined

Configured for VC-WIN32.

c:\relax\openssl-1.0.2m>call ms\do_nasm

c:\relax\openssl-1.0.2m>perl util\mkfiles.pl  1>MINFO

c:\relax\openssl-1.0.2m>perl util\mk1mf.pl nasm VC-WIN32  1>ms\nt.mak

c:\relax\openssl-1.0.2m>perl util\mk1mf.pl dll nasm VC-WIN32  1>ms\ntdll.mak

c:\relax\openssl-1.0.2m>perl util\mk1mf.pl nasm BC-NT  1>ms\bcb.mak

c:\relax\openssl-1.0.2m>perl util\mkdef.pl 32 libeay  1>ms\libeay32.def

c:\relax\openssl-1.0.2m>perl util\mkdef.pl 32 ssleay  1>ms\ssleay32.def

c:\relax\openssl-1.0.2m>nmake -f ms\nt.mak

Microsoft (R) Program Maintenance Utility Version 10.00.30319.01
Copyright (C) Microsoft Corporation.  All rights reserved.

Building OpenSSL
        perl util/mkdir-p.pl "tmp32"
created directory `tmp32'
        perl util/mkdir-p.pl "out32"
created directory `out32'
        perl util/mkdir-p.pl "inc32"
created directory `inc32'
        perl util/mkdir-p.pl "inc32\openssl"
created directory `inc32/openssl'
        perl .\util\copy-if-different.pl ".\.\e_os.h" "tmp32\e_os.h"
Can't Open "././e_os.h" at .\util\copy-if-different.pl line 67.
NMAKE : fatal error U1077: 'c:\relax\strawberry\perl\bin\perl.EXE' : return code '0x16'
Stop.

c:\relax\openssl-1.0.2m>nmake -f ms\nt.mak test

Microsoft (R) Program Maintenance Utility Version 10.00.30319.01
Copyright (C) Microsoft Corporation.  All rights reserved.

        cl /Fotmp32\constant_time_test.obj -Iinc32 -Itmp32 /MT /Ox /O2 /Ob2 -DOPENSSL_THREADS  -DDSO_WIN32 -W3 -Gs0 -GF -Gy -nologo -DOPENSSL_SYSNAME_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -D_CRT_SECU
RE_NO_DEPRECATE -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM
-DWHIRLPOOL_ASM -DGHASH_ASM -DOPENSSL_NO_RC5 -DOPENSSL_NO_MD2 -DOPENSSL_NO_SSL2 -DOPENSSL_NO_KRB5 -DOPENSSL_NO_JPAKE -DOPENSSL_NO_WEAK_SSL_CIPHERS -DOPENSSL_NO_DYNAMIC_ENGINE /Zi /Fdtmp32/app -c .\cry
pto\constant_time_test.c
constant_time_test.c
c:\relax\openssl-1.0.2m\crypto\../crypto/constant_time_locl.h(50) : fatal error C1083: Cannot open include file: 'e_os.h': No such file or directory
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin\cl.EXE"' : return code '0x2'
Stop.

c:\relax\openssl-1.0.2m>nmake -f ms\nt.mak install

Microsoft (R) Program Maintenance Utility Version 10.00.30319.01
Copyright (C) Microsoft Corporation.  All rights reserved.

Building OpenSSL
        perl .\util\copy-if-different.pl ".\.\e_os.h" "tmp32\e_os.h"
Can't Open "././e_os.h" at .\util\copy-if-different.pl line 67.
NMAKE : fatal error U1077: 'c:\relax\strawberry\perl\bin\perl.EXE' : return code '0x16'
Stop.

c:\relax\openssl-1.0.2m>popd

C:\relax>endlocal
C:\relax>
```

** many thanks for maintaining this over so many years **

**Comment by mattcaswell** [comment]
> Can't Open "././e_os.h" at .\util\copy-if-different.pl line 67.

That is quite strange. The file "e_os.h" should just be in the top level directory of the unpacked files. Is it there? Is it readable?

**Comment by dch** [comment]
I just confirmed that this issue is also present in latest snapshot `openssl-1.0.2-stable-SNAP-20171106` too.

@mattcaswell thanks. Yes, if I add a few `s!^\.//\.//!!g;` just prior, this gets me a bit further. If I use a simple `system("copy","/b","y",$_,$dfile)` then this also gets me further, but I still run into similar issues later on.

the files are definitely there and readable, its a local filesystem, I'm running the build as administrator.

**Comment by dch** [comment]
My perl version is strawberry perl v5.12.2 FWIW, reasonably old, I'll re-try with a newer version v5.26.1.

**Comment by mattcaswell** [comment]
Please do that first and let us know the results. Looking at the "copy-if-different" script there has only been one recent change (the one before that was over 4 years ago). That was in commit b83265697 via PR #3904. To revert the commit that made that change:

````
diff --git a/util/copy-if-different.pl b/util/copy-if-different.pl
index e1245f54af..ec99e084b5 100644
--- a/util/copy-if-different.pl
+++ b/util/copy-if-different.pl
@@ -12,7 +12,7 @@ my @filelist;
 
 foreach my $arg (@ARGV) {
        $arg =~ s|\\|/|g;       # compensate for bug/feature in cygwin glob...
-       foreach (glob qq("$arg"))
+       foreach (glob $arg)
                {
                push @filelist, $_;
                }
diff --git a/util/copy.pl b/util/copy.pl
index a6b2a54ec6..eba6d5815e 100644
--- a/util/copy.pl
+++ b/util/copy.pl
@@ -19,7 +19,7 @@ foreach $arg (@ARGV) {
                next;
                }
        $arg =~ s|\\|/|g;       # compensate for bug/feature in cygwin glob...
-       foreach (glob qq("$arg"))
+       foreach (glob $arg)
                {
                push @filelist, $_;
                }
````

**Comment by dch** [comment]
\o/ updated perl does indeed make this go away. closing with prejudice... sorry for the noise

**Comment by dot-asm** [comment]
Since we don't specify minimum Perl version that high(*) the problem was addressed in #4696.

(*) Requirement for 1.1.x is 5.10, requirement for 1.0.2 de-facto can't be higher.
